framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1109,53 +1188,134 @@ int pcm_stop(struct pcm *pcm)
 		return -EINVAL;
 	}
 #ifdef CONFIG_AUDIO_MULTI_SESSION
-	if (ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
+	if (!pcm->running && ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
 #else
-	if (ioctl(pcm->fd, AUDIOIOC_STOP, 0) < 0)
+	if (!pcm->running && ioctl(pcm->fd, AUDIOIOC_STOP, 0) < 0)
 #endif
 		return oops(pcm, errno, "cannot stop channel");
 
+	/* Remove any pending messages from the message queue */
+	struct timespec st_time;
+	struct audio_msg_s msg;
+	unsigned int size;
+	int prio;
+	while (1) {
+		clock_gettime(CLOCK_REALTIME, &st_time);
+		size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
+		if (size != sizeof(msg)) {
+			break;
+		}
+	}
+
 	pcm->prepared = 0;
 	pcm->running = 0;
-	pcm->bufPtr = 0;
-	pcm->nextSize = 0;
-	pcm->nextOffset = 0;
-	pcm->nextBuf = NULL;
+	pcm->draining = 0;
+	pcm->buf_ptr = 0;
+	pcm->next_size = 0;
+	pcm->next_offset = 0;
+	pcm->next_buf = NULL;
 
 	return 0;
 }
 
+/** Stops a PCM. Any data present in the buffers will be dropped.
+ * @param pcm A PCM handle.
+ * @return On success, zero; on failure, a negative number.
+ * @ingroup libtinyalsa-pcm
+ */
+int pcm_drop(struct pcm *pcm)
+{
+	return pcm_stop(pcm);
+}
+
+/** Stop a PCM preserving pending frames.
+ * @param pcm A PCM handle.
+ * @return On success, zero; on failure, a negative number.
+ * @ingroup libtinyalsa-pcm
+ */
+int pcm_drain(struct pcm *pcm)
+{
+	if (pcm == NULL) {
+		return -EINVAL;
+	}
+
+	if (!pcm->running) {
+		return oops(pcm, EINVAL, "PCM is already stopped.");
+	}
+
+	if (pcm->flags & PCM_OUT) {
```

Comment:
When playback, pcm_drain should play remained data in buffer, but without new ioctl command, is it possibile?

Reply:
Yes
When you call pcm_drain during playback, it will flush all data in buffer and then stop the playback. 
There is no need for a new ioctl.

Reply:
Below is generic usage, below works with your code? because I can't see request play with remained data in pcm_drain. If it works, please let me know.
for() {
    read(fd, buf, size);
    ret = pcm_writei();
    if (ret == -EPIPE) {
         pcm_prepare();
    } else if (ret < 0) {
        oops();
   }
}
pcm_drain();
pcm_close();

Reply:
Yes. It will work. 

for(){
pcm_write() --->> here play is started. Kernel will continue to play till it has any buffer in its queue
...
...
}
pcm_drain();
   - wait in a while loop for deque messages
   - each deque msg indicates that kernel has played one buffer
   - block untill all buffers are played
   - pcm_stop()
pcm_close()